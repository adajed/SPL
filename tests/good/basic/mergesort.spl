void mergesort(int[] arr, int s, int e)
{
    if (s >= e - 1)
        return;

    int m = (s + e) / 2;

    mergesort(arr, s, m);
    mergesort(arr, m, e);

    int[] newarr = new int[e - s];

    int i = s, j = m, p = 0;
    while (i < m || j < e)
    {
        if (i == m)
        {
            newarr[p] = arr[j];
            j++;
        }
        else if (j == e)
        {
            newarr[p] = arr[i];
            i++;
        }
        else
        {
            if (arr[i] < arr[j])
            {
                newarr[p] = arr[i];
                i++;
            }
            else
            {
                newarr[p] = arr[j];
                j++;
            }
        }

        p++;
    }

    i = s;
    p = 0;
    while (i < e)
    {
        arr[i] = newarr[p];
        i++;
        p++;
    }

    return;
}

void printArray(int[] arr)
{
    for (i = 0 to arr.length)
        printInt(arr[i]);

    return;
}

int main()
{
    int n = 10;
    int[] arr = new int[n];

    int b = 1;
    for (i = 0 to arr.length)
    {
        arr[i] = b * i * i;
        b = -b;
    }

    printArray(arr);
    mergesort(arr, 0, n);
    printArray(arr);

    return 0;
}
