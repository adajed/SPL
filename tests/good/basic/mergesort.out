0
-1
4
-9
16
-25
36
-49
64
-81
-81
-49
-25
-9
-1
0
4
16
36
64
